daemon/mounts.go

```diff
@@ -38,28 +38,49 @@ func (daemon *Daemon) removeMountPoints(container *container.Container, rm bool)
 	var rmErrors []string
 	ctx := context.TODO()
 	for _, m := range container.MountPoints {
-		if m.Type != mounttypes.TypeVolume || m.Volume == nil {
-			continue
-		}
-		daemon.volumes.Release(ctx, m.Volume.Name(), container.ID)
-		if !rm {
-			continue
-		}
+		if m.Type == mounttypes.TypeVolume {
+			if m.Volume == nil {
+				continue
+			}
+			daemon.volumes.Release(ctx, m.Volume.Name(), container.ID)
+			if !rm {
+				continue
+			}
 
-		// Do not remove named mountpoints
-		// these are mountpoints specified like `docker run -v <name>:/foo`
-		if m.Spec.Source != "" {
-			continue
+			// Do not remove named mountpoints
+			// these are mountpoints specified like `docker run -v <name>:/foo`
+			if m.Spec.Source != "" {
+				continue
+			}
+
+			err := daemon.volumes.Remove(ctx, m.Volume.Name())
+			// Ignore volume in use errors because having this
+			// volume being referenced by other container is
+			// not an error, but an implementation detail.
+			// This prevents docker from logging "ERROR: Volume in use"
+			// where there is another container using the volume.
+			if err != nil && !volumesservice.IsInUse(err) {
+				rmErrors = append(rmErrors, err.Error())
+			}
 		}
 
-		err := daemon.volumes.Remove(ctx, m.Volume.Name())
-		// Ignore volume in use errors because having this
-		// volume being referenced by other container is
-		// not an error, but an implementation detail.
-		// This prevents docker from logging "ERROR: Volume in use"
-		// where there is another container using the volume.
-		if err != nil && !volumesservice.IsInUse(err) {
-			rmErrors = append(rmErrors, err.Error())
+		if m.Type == mounttypes.TypeImage {
+			layerName := fmt.Sprintf("%s-%s", container.ID, m.Spec.Source)
+			layer, err := daemon.imageService.GetLayerByID(layerName)
+			if err != nil {
+				rmErrors = append(rmErrors, err.Error())
+				continue
+			}
+			err = daemon.imageService.ReleaseLayer(layer)
+			if err != nil {
+				rmErrors = append(rmErrors, err.Error())
+				continue
+			}
+			err = layer.Unmount()
+			if err != nil {
+				rmErrors = append(rmErrors, err.Error())
+				continue
+			}
```

Comment:
The graphdriver implementation of `GetLayerByID` will actually create a new reference to that layer, so the this won't free the layer created with `CreateLayerFromImage`.

We should probably store it the same way we do with `RWLayer`.

Alternatively, we could initially drop the graphdriver implementation and only implement it for containerd store.

Reply:
I would rather not drop the graph driver because lots of users would benefit this out of the gate and o would prefer not having to switch to containerd as a limit.

Reply:
Fixed!